services/datamanager/builtin/builtin.go

```diff
@@ -697,62 +697,85 @@ func isOffline() bool {
 
 func (svc *builtIn) sync() {
 	svc.flushCollectors()
-
+	// Lock while retrieving any values that could be changed during reconfiguration of the data
+	// manager.
 	svc.lock.Lock()
-	if svc.syncer != nil {
-		toSync := getAllFilesToSync(svc.captureDir, svc.fileLastModifiedMillis)
-		for _, ap := range svc.additionalSyncPaths {
-			toSync = append(toSync, getAllFilesToSync(ap, svc.fileLastModifiedMillis)...)
-		}
-		syncer := svc.syncer
-		stopAfter := time.Now().Add(time.Duration(svc.syncIntervalMins * float64(time.Minute)))
-		svc.lock.Unlock()
-
-		// Only log if there are a large number of files to sync
-		if len(toSync) > minNumFiles {
-			svc.logger.Infof("Starting sync of %d files", len(toSync))
-		}
-		for _, p := range toSync {
-			syncer.SyncFile(p, stopAfter)
-		}
-	} else {
+	captureDir := svc.captureDir
+	fileLastModifiedMillis := svc.fileLastModifiedMillis
+	additionalSyncPaths := svc.additionalSyncPaths
+	if svc.syncer == nil {
 		svc.lock.Unlock()
+		return
 	}
+	syncer := svc.syncer
+	svc.lock.Unlock()
+
+	// Kick off a goroutine to retrieve all the names of the files to sync, then another 1000 to
+	// sync the files to Viam app.
+
+	svc.backgroundWorkers.Add(1)
+	go func() {
```

Comment:
I'm not sure if there's a name for this, but I'm using the phrase "one sync job" or "a sync round" to describe all of the work generated from a single "sync interval".

From my reading, this goroutine usage means that if there's a network error, we'll start accumulating "sync job goroutines" that, once the network comes back, can overlap.

Generally (pre-this-patch), I think overlapping is confusing. But with this PR, these goroutines will just get blocked on adding (many similar) files to the `syncer.filesToSync` channel.

I'd like to see us not use a goroutine here. That way we won't overlap "jobs" and have to worry about whether that's a contributing factor to memory accumulating over time.

And longer term if there's a need to allow for overlapping, we can be a bit more thoughtful about doing it safely.

Reply:
Talked with dan offline, implication is to call getAllFilesToSync not in its in go routines as `sync()` is called from a different goroutine. @agreenb Does that change make sense?